hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,409 @@
+# %%
+import warnings
+from collections import Counter, defaultdict
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from .._ipcw import KaplanMeierIPCW
+from ..utils import check_y_survival
+
+
+def concordance_index_incidence(
+    y_test,
+    y_pred,
+    time_grid=None,
+    taus=None,
+    y_train=None,
+    X_train=None,
+    X_test=None,
+    event_of_interest=1,
+    ipcw_estimator="km",
+    tied_tol=1e-8,
+):
+    """Time-dependent concordance index for prognostic models with competing risks \
+    using inverse probability of censoring weighting.
+
+    The concordance index (C-index) is a common metric in survival analysis that
+    evaluates whether the model predictions correctly order pairs of individuals with
+    respect to the timing of the event of interest. It is defined as the probability
+    that the prediction is concordant for a pair of individuals and is computed as the
+    ratio of the number of concordant pairs to the total number of pairs.
+    This implementation extends the C-index to the competing risks setting, where
+    multiple alternative events are considered, aiming to determine which event occurs
+    first and when, following the formulas and notations in [1]_.
+
+    Due to the right-censoring in the data, the order of some pairs is unknown,
+    so we define the notion of comparable pairs, i.e. the pairs for which
+    we can compare the order of occurrence of the event of interest.
+    A pair :math:`(i, j)` is comparable, with :math:`i` experiencing the event of
+    interest at time :math:`T_i` if:
+
+    - :math:`j` experiences the event of interest at a strictly greater time
+      :math:`T_j > T_i` (pair of type A)
+    - :math:`j` is censored at time :math:`T_j = T_i` or greater (pair of type A)
+    - :math:`j` experiences a competing event before or at time :math:`T_i`
+      (pair of type B)
+
+    The pair :math:`(i, j)` is considered a tie for time if :math:`j` experiences
+    the event of interest at the same time (:math:`T_j=T_i`). This tied time pair will
+    be counted as :math:`1/2` for the count of comparable pairs.
```

Comment:
This tie in time pair is not considered in ref [1] ?

Reply:
I think the type A with T_i=T_j is not consistent with ref [1] (where type A is T_i < T_j , and any event for j). Maybe you have a mix of ref [1] and `sksurv.metrics.concordance_index_ipcw` conventions ?

Are you maybe adding events (T_i=T_j and D_i=D_j=event_of_interest), discarded in ref [1], but here counted as "tied in time" with 1/2 weight ?

Reply:
Yes, you are right, we chose to include them to be closer to the results of `sksurv` in the survival analysis setting (i.e. no competing events). Maybe @judithabk6 can add a little bit more context for our understanding?

Reply:
Actually, let's remove this section from the docstring because we don't use the ties in times. We only use the ties in predictions